### PR TITLE
runtimetest: Make TAP output more granular

### PR DESCRIPTION
**Pull request by wking**
Granular TAP output makes it easier to see what's being tested (and what's going wrong).  You could feed this into any TAP harness you like, but I've chosen prove (it seems popular).  You need a TAP harness looking for test failures, because runtimetest now returns zero when it successfully runs the tests (regardless of whether the tests all pass).  To avoid confusing the TAP consumer, all of the non-TAP logging goes to stderr (and not stdout).

With this change, the verbose output looks like:

    # prove -v ./test_runtime.sh :: -r ~/bin/runc
    ./test_runtime.sh ..
    TAP version 13
    ok 1 # SKIP root.readonly falsy
    ok 2 - hostname matches expected value
    ok 3 - mount "/proc" found
    ok 4 - mount "/dev" found
    ok 5 - mount "/dev/pts" found
    ok 6 - mount "/dev/shm" found
    ok 7 - mount "/dev/mqueue" found
    ok 8 - mount "/sys" found
    ok 9 - expected capability CAP_CHOWN set
    ok 10 - expected capability CAP_DAC_OVERRIDE set
    ok 11 - unexpected capability CAP_DAC_READ_SEARCH not set
    ok 12 - expected capability CAP_FOWNER set
    ok 13 - expected capability CAP_FSETID set
    ok 14 - expected capability CAP_KILL set
    ok 15 - expected capability CAP_SETGID set
    ok 16 - expected capability CAP_SETUID set
    ok 17 - expected capability CAP_SETPCAP set
    ok 18 - unexpected capability CAP_LINUX_IMMUTABLE not set
    ok 19 - expected capability CAP_NET_BIND_SERVICE set
    ok 20 - unexpected capability CAP_NET_BROADCAST not set
    ok 21 - unexpected capability CAP_NET_ADMIN not set
    ok 22 - expected capability CAP_NET_RAW set
    ok 23 - unexpected capability CAP_IPC_LOCK not set
    ok 24 - unexpected capability CAP_IPC_OWNER not set
    ok 25 - unexpected capability CAP_SYS_MODULE not set
    ok 26 - unexpected capability CAP_SYS_RAWIO not set
    ok 27 - expected capability CAP_SYS_CHROOT set
    ok 28 - unexpected capability CAP_SYS_PTRACE not set
    ok 29 - unexpected capability CAP_SYS_PACCT not set
    ok 30 - unexpected capability CAP_SYS_ADMIN not set
    ok 31 - unexpected capability CAP_SYS_BOOT not set
    ok 32 - unexpected capability CAP_SYS_NICE not set
    ok 33 - unexpected capability CAP_SYS_RESOURCE not set
    ok 34 - unexpected capability CAP_SYS_TIME not set
    ok 35 - unexpected capability CAP_SYS_TTY_CONFIG not set
    ok 36 - expected capability CAP_MKNOD set
    ok 37 - unexpected capability CAP_LEASE not set
    ok 38 - expected capability CAP_AUDIT_WRITE set
    ok 39 - unexpected capability CAP_AUDIT_CONTROL not set
    ok 40 - expected capability CAP_SETFCAP set
    ok 41 - unexpected capability CAP_MAC_OVERRIDE not set
    ok 42 - unexpected capability CAP_MAC_ADMIN not set
    ok 43 - unexpected capability CAP_SYSLOG not set
    ok 44 - unexpected capability CAP_WAKE_ALARM not set
    ok 45 - unexpected capability CAP_BLOCK_SUSPEND not set
    ok 46 - unexpected capability CAP_AUDIT_READ not set
    ok 47 - lstat default symlink /dev/fd
    ok 48 - default symlink /dev/fd is a symlink
    ok 49 - default symlink /dev/fd has expected target
    ok 50 - lstat default symlink /dev/stdin
    ok 51 - default symlink /dev/stdin is a symlink
    ok 52 - default symlink /dev/stdin has expected target
    ok 53 - lstat default symlink /dev/stdout
    ok 54 - default symlink /dev/stdout is a symlink
    ok 55 - default symlink /dev/stdout has expected target
    ok 56 - lstat default symlink /dev/stderr
    ok 57 - default symlink /dev/stderr is a symlink
    ok 58 - default symlink /dev/stderr has expected target
    ok 59 - mount /proc has expected type
    ok 60 - mount /sys has expected type
    ok 61 - mount /dev/pts has expected type
    ok 62 - mount /dev/shm has expected type
    ok 63 - stat default device /dev/null
    ok 64 - default device /dev/null is a device
    ok 65 - stat default device /dev/zero
    ok 66 - default device /dev/zero is a device
    ok 67 - stat default device /dev/full
    ok 68 - default device /dev/full is a device
    ok 69 - stat default device /dev/random
    ok 70 - default device /dev/random is a device
    ok 71 - stat default device /dev/urandom
    ok 72 - default device /dev/urandom is a device
    ok 73 - stat default device /dev/tty
    ok 74 - default device /dev/tty is a device
    ok 75 - stat default device /dev/ptmx
    ok 76 - default device /dev/ptmx is a device
    ok 77 # SKIP linux.devices (no devices configured)
    ok 78 - has expected working directory
    ok 79 - has expected environment variable PATH
    ok 80 - has expected environment variable TERM
    ok 81 - has expected user ID
    ok 82 - has expected group ID
    ok 83 - has expected number of process arguments
    ok 84 - has expected process argument 0
    ok 85 - has expected noNewPrivileges
    ok 86 # SKIP linux.resources.oomScoreAdj falsy
    ok 87 - has expected soft RLIMIT_NOFILE
    ok 88 - has expected hard RLIMIT_NOFILE
    ok 89 # SKIP linux.uidMappings checks (no mappings specified)
    ok 90 # SKIP linux.gidMappings checks (no mappings specified)
    1..90
    ok
    All tests successful.
    Files=1, Tests=90,  0 wallclock secs ( 0.02 usr  0.00 sys +  0.04 cusr  0.01 csys =  0.07 CPU)
    Result: PASS

### CODE REVIEW

**Comment by Mashimiao** [comment]
I think that's too detailed...

**Comment by wking** [comment]
On Tue, Jan 17, 2017 at 12:22:00AM -0800, Ma Shimiao wrote:
> I think that's too detailed...

Most users will not consume the raw TAP.  They'll be looking at the
prove summary [1] (which is only a few lines).  The detail is just for
the curious and for failures.  For an example of the failure display,
here's what you get with a no-op runtime:

  $ ./oci-runtime-tool generate >config.json
  $ prove ./runtimetest
  ./runtimetest .. Failed 22/90 subtests
          (less 5 skipped subtests: 63 okay)

  Test Summary Report
  -------------------
  ./runtimetest (Wstat: 0 Tests: 90 Failed: 22)
    Failed tests:  2, 4, 9-10, 12-17, 19, 22, 27, 36, 38, 40
                  78-79, 81-82, 84, 88
  Files=1, Tests=90,  0 wallclock secs ( 0.02 usr +  0.00 sys =  0.02 CPU)
  Result: FAIL

Or showing the failed tests:

  $ prove -f ./runtimetest
  ./runtimetest .. 1/?
  not ok 2 - hostname matches expected value
  not ok 4 - expected mount "/dev" found
  not ok 9 - expected capability CAP_CHOWN set
  not ok 10 - expected capability CAP_DAC_OVERRIDE set
  not ok 12 - expected capability CAP_FOWNER set
  not ok 13 - expected capability CAP_FSETID set
  not ok 14 - expected capability CAP_KILL set
  not ok 15 - expected capability CAP_SETGID set
  not ok 16 - expected capability CAP_SETUID set
  not ok 17 - expected capability CAP_SETPCAP set
  not ok 19 - expected capability CAP_NET_BIND_SERVICE set
  not ok 22 - expected capability CAP_NET_RAW set
  not ok 27 - expected capability CAP_SYS_CHROOT set
  not ok 36 - expected capability CAP_MKNOD set
  not ok 38 - expected capability CAP_AUDIT_WRITE set
  not ok 40 - expected capability CAP_SETFCAP set
  not ok 78 - has expected working directory
  not ok 79 - has expected environment variable PATH
  not ok 81 - has expected user ID
  not ok 82 - has expected group ID
  not ok 84 - has expected process argument 0
  not ok 88 - has expected hard RLIMIT_NOFILE
  ./runtimetest .. Failed 22/90 subtests
          (less 5 skipped subtests: 63 okay)

  Test Summary Report
  -------------------
  ./runtimetest (Wstat: 0 Tests: 90 Failed: 22)
    Failed tests:  2, 4, 9-10, 12-17, 19, 22, 27, 36, 38, 40
                  78-79, 81-82, 84, 88
  Files=1, Tests=90,  0 wallclock secs ( 0.02 usr +  0.00 sys =  0.02 CPU)
  Result: FAIL

Or showing the failures and comments:

  $ prove -fo ./runtimetest
  ./runtimetest .. 1/?
  not ok 2 - hostname matches expected value
  # hostname expected: mrsdalloway, actual: my-hostname
  not ok 4 - expected mount "/dev" found
  not ok 9 - expected capability CAP_CHOWN set
  not ok 10 - expected capability CAP_DAC_OVERRIDE set
  not ok 12 - expected capability CAP_FOWNER set
  not ok 13 - expected capability CAP_FSETID set
  not ok 14 - expected capability CAP_KILL set
  not ok 15 - expected capability CAP_SETGID set
  not ok 16 - expected capability CAP_SETUID set
  not ok 17 - expected capability CAP_SETPCAP set
  not ok 19 - expected capability CAP_NET_BIND_SERVICE set
  not ok 22 - expected capability CAP_NET_RAW set
  not ok 27 - expected capability CAP_SYS_CHROOT set
  not ok 36 - expected capability CAP_MKNOD set
  not ok 38 - expected capability CAP_AUDIT_WRITE set
  not ok 40 - expected capability CAP_SETFCAP set
  not ok 78 - has expected working directory
  # working directory expected: /, actual: /home/wking/src/runtime-tools
  not ok 79 - has expected environment variable PATH
  # environment variable PATH expected: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin, actual: /home/wking/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/sbin:/sbin
  not ok 81 - has expected user ID
  # user ID expected: 0, actual: 1000
  not ok 82 - has expected group ID
  # group ID expected: 0, actual: 1000
  not ok 84 - has expected process argument 0
  # expected process argument 0: sh, actual: init [3]
  not ok 88 - has expected hard RLIMIT_NOFILE
  # hard RLIMIT_NOFILE expected: 1024, actual: 4096
  ./runtimetest .. Failed 22/90 subtests
          (less 5 skipped subtests: 63 okay)

  Test Summary Report
  -------------------
  ./runtimetest (Wstat: 0 Tests: 90 Failed: 22)
    Failed tests:  2, 4, 9-10, 12-17, 19, 22, 27, 36, 38, 40
                  78-79, 81-82, 84, 88
  Files=1, Tests=90,  0 wallclock secs ( 0.02 usr +  0.00 sys =  0.02 CPU)
  Result: FAIL

So you can easily tune the output to be as noisy as you like.

[1]: https://github.com/wking/ocitools-v2/blob/459c3848b94184d4831a00f4bff47b79592352b7/README.md#testing-oci-runtimes

**Comment by mrunalp** [comment]
My first reaction was that it was too granular but I see how it could be helpful.

**Comment by Mashimiao** [comment]
@wking   can you rebase this?

**Comment by wking** [comment]
On Tue, Mar 07, 2017 at 09:36:49PM -0800, Ma Shimiao wrote:
> @wking   can you rebase this?

I expect this is going to be a pain in the neck to continually rebase
;).  Can we at least get a nominal ok on the general idea from another
maintainer before I go through the trouble?

**Comment by Mashimiao** [comment]
@liangchenye   How do you think about this PR?

**Comment by wking** [comment]
If folks are considering [v0.4.0][1] PRs in [ascending-date order][2], I'd prefer they skip over this one and look at #439 first.  That PR is current, and it's easier to rebase than this one.  I'll try and get this rebased in the next week or so, but this PR will need another rebase if the older and also milestoned for v0.4.0 #93 lands first.

[1]: https://github.com/opencontainers/runtime-tools/milestone/6
[2]: https://github.com/opencontainers/runtime-tools/pull/439#issuecomment-346752891

**Comment by Mashimiao** [comment]
#93  needs more work, I have put it into lower priority than this one.

**Comment by q384566678** [comment]
@wking Others have been merged, I think you can update this pr.

**Comment by wking** [comment]
On Thu, Jan 11, 2018 at 02:00:54AM +0000, Zhou Hao wrote:
> @wking Others have been merged, I think you can update this pr.

I've started in on that, but it's going to take me a few days to catch
things up.  I've filed #552 to slightly reduce the number of places I
have to edit in this PR.

**Comment by q384566678** [comment]
@wking Excuse me, how about this patch update? I'm ready release 0.4 version, I think this patch need to merge before the release.

**Comment by wking** [comment]
Sorry, I'm only ~ half way through, and I still need to make adjustments to handle `specerror`.  Maybe bump this to 0.5?

**Comment by q384566678** [comment]
Fine.

**Comment by q384566678** [comment]
@wking Can you update this pr? So we can put it in 0.6.

**Comment by wking** [comment]
Finally rebased onto master with 459c384 → 7c384c1.  The diff is huge, and I only just got through it, so give it a careful read before merging.  On the other hand, don't wait too long or we'll have another big rebase to go through ;).  I'll go back through in the next few days and try to double check things as well.

File: cmd/runtimetest/main.go [hunk 1]
```diff
@@ -278,63 +354,148 @@ func validateRlimits(spec *rspec.Spec, t *tap.T) error {
...
+	if err != nil {
+		return false, nil
+	}
+	defer f.Close()
+	b := make([]byte, 1)
+	_, err = f.Read(b)
+	if err == nil {
+		return true, nil
+	} else if err == io.EOF {
+		return false, nil
```
Comment (alban) [inline cmd/runtimetest/main.go]: I find this surprising to return "non-readable" when the file might well be readable but just empty.

File: cmd/runtimetest/main.go [hunk 2]
```diff
@@ -278,63 +354,148 @@ func validateRlimits(spec *rspec.Spec, t *tap.T) error {
...
+}
+
+func testDirectoryWriteAccess(path string) (writable bool, err error) {
+	tmpfile, err := ioutil.TempFile(path, "Test")
+	if err != nil {
+		return false, nil
+	}
 	tmpfile.Close()
-	os.RemoveAll(filepath.Join(path, tmpfile.Name()))
+	if err != nil {
```
Comment (alban) [inline cmd/runtimetest/main.go]: this branch cannot be taken because it was tested a few lines above.

File: cmd/runtimetest/main.go (hunk 1 above)
Comment (wking) [reply cmd/runtimetest/main.go]: > I find this surprising to return "non-readable" when the file might well be readable but just empty.

Yeah, but I'm not sure how to detect readable-but-empty (and what I have here is what we've [been using][1]).  Do you have ideas?  The lack of support for that case makes this brittle, but I'm ok with it if we update our tests (e.g. [here][2]) to use non-empty test files.  `runtimetest` is for internal testing only (not part of our public API), so a few cludges to work around the `runtimetest` limitation are unfortunate but survivable.

[1]: https://github.com/opencontainers/runtime-tools/blob/fcd5b23b0f05aa34cb922fe39c6ba79e6f46fe39/cmd/runtimetest/main.go#L534-L543
[2]: https://github.com/opencontainers/runtime-tools/blob/fcd5b23b0f05aa34cb922fe39c6ba79e6f46fe39/validation/linux_readonly_paths.go#L14-L15

File: cmd/runtimetest/main.go (hunk 2 above)
Comment (wking) [reply cmd/runtimetest/main.go]: > this branch cannot be taken because it was tested a few lines above.

Thanks :).  I've dropped it with d7e4d91 → 857249d.

File: cmd/runtimetest/main.go (hunk 1 above)
Comment (wking) [reply cmd/runtimetest/main.go]: Actually, our current tests use an empty file and [treat an immediate EOF as success][1].  So I am changing the logic.  But the old logic doesn't make much sense to me.  What if the runtime implemented `linux.readonlyPaths` by binding `/dev/null` over the listed paths?  That would seem to satisfy [the spec][2] (which is pretty weak on this anyway), while still resulting in immediate EOFs.

[1]: https://github.com/opencontainers/runtime-tools/blob/fcd5b23b0f05aa34cb922fe39c6ba79e6f46fe39/cmd/runtimetest/main.go#L540-L542
[2]: https://github.com/opencontainers/runtime-spec/blob/v1.0.1/config-linux.md#readonly-paths

Comment (alban) [reply cmd/runtimetest/main.go]: Maybe just add a comment to explain the situation:
```
	} else if err == io.EOF {
                // linux_readonly_paths.go only tests non-empty files. So if we find
                // an empty file, the runtime did not apply the ReadonlyPaths correctly
		return false, nil
```

I agree about updating the test you mention :)

File: cmd/runtimetest/main.go [hunk 3]
```diff
@@ -372,36 +530,46 @@ func validateRootfsPropagation(spec *rspec.Spec, t *tap.T) error {
...
+			exposed = false
+		} else if err != nil {
+			return err
+		} else {
+			exposed = true
 		}
 		if spec.Linux.RootfsPropagation == "shared" {
-			return nil
+			c.harness.Ok(exposed, fmt.Sprintf("shared root propogation exposes %q", targetFile))
+		} else {
```
Comment (alban) [inline cmd/runtimetest/main.go]: I find the [spec](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#rootfs-mount-propagation) quite weak about this RootfsPropagation:
- a mount could be both shared and slave
- `rootfsPropagation=slave` is only meaningful if we say what is the master of that enslaved rootfs. Should it necessarily be wrt to a mount in the host namespace?

But for the purpose of this PR, your change is fine.

File: cmd/runtimetest/main.go [hunk 4]
```diff
@@ -410,201 +578,379 @@ func validateDefaultFS(spec *rspec.Spec, t *tap.T) error {
...
+	var exists bool
+	fi, err := os.Stat(device.Path)
+	if os.IsNotExist(err) {
+		exists = false
+	} else if err != nil {
+		return err
+	} else {
+		exists = true
+	}
+	rfcError, err := c.Ok(exists, specerror.DevicesAvailable, version, fmt.Sprintf("has a file at %s", description))
```
Comment (alban) [inline cmd/runtimetest/main.go]: `specerror.DevicesAvailable` -> `condition`

Comment (alban) [reply cmd/runtimetest/main.go]: ditto in the 6 other occurrences of `specerror.DevicesAvailable` below in this function

File: cmd/runtimetest/main.go [hunk 5]
```diff
@@ -410,201 +578,379 @@ func validateDefaultFS(spec *rspec.Spec, t *tap.T) error {
...
+		c.harness.YAML(map[string]interface{}{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"path":      device.Path,
+			"expected":  expectedPerm,
+			"actual":    actualPerm,
+		})
+	}
+
+	if description == "/dev/console (default device)" {
```
Comment (alban) [inline cmd/runtimetest/main.go]: wouldn't it be better to test
```
if device.Path == "/dev/console" {
```

File: cmd/runtimetest/main.go [hunk 6]
```diff
@@ -410,201 +578,379 @@ func validateDefaultFS(spec *rspec.Spec, t *tap.T) error {
...
+		c.harness.YAML(map[string]string{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"path":      symlink,
+		})
+		if !exists {
+			continue
+		}
+
+		isSymlink := fi.Mode()&os.ModeSymlink == os.ModeSymlink
```
Comment (alban) [inline cmd/runtimetest/main.go]: I find it more clear written this way:
```
isSymlink := fi.Mode()&os.ModeType == os.ModeSymlink
```
but that's equivalent in Golang because os.ModeType is a bitfield.

It confused me at first because in C, `stat.st_mode` is not a bitfield (man 7 inode):
`S_IFLNK = S_IFREG | S_IFCHR`
```
       The following mask values are defined for the file type:

           S_IFMT     0170000   bit mask for the file type bit field

           S_IFSOCK   0140000   socket
           S_IFLNK    0120000   symbolic link
           S_IFREG    0100000   regular file
           S_IFBLK    0060000   block device
           S_IFDIR    0040000   directory
           S_IFCHR    0020000   character device
           S_IFIFO    0010000   FIFO
```

(Things I learnt today)

Comment (alban) [reply cmd/runtimetest/main.go]: I see it was already written like that before. Feel free to ignore my comment ;)

File: cmd/runtimetest/main.go [hunk 7]
```diff
@@ -410,201 +578,379 @@ func validateDefaultFS(spec *rspec.Spec, t *tap.T) error {
...
 		return nil
 	}
-	for _, v := range spec.Linux.ReadonlyPaths {
-		err := testWriteAccess(v)
-		if err == nil {
-			return fmt.Errorf("%v should be readonly", v)
+
+	for i, path := range spec.Linux.ReadonlyPaths {
+		readable, err := testReadAccess(path)
+		if err != nil && !os.IsNotExist(err) {
```
Comment (alban) [inline cmd/runtimetest/main.go]: Why ignoring the `os.IsNotExist` error?

File: cmd/runtimetest/main.go [hunk 8]
```diff
@@ -410,201 +578,379 @@ func validateDefaultFS(spec *rspec.Spec, t *tap.T) error {
...
+		}
+		if !readable {
+			c.harness.Skip(1, fmt.Sprintf("%q (linux.readonlyPaths[%d]) is not readable", path, i))
+		}
+
+		writable, err := testWriteAccess(path)
+		if err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		if !writable {
```
Comment (alban) [inline cmd/runtimetest/main.go]: why the `if`? The condition is already in the `c.harness.Ok()` below.

File: cmd/runtimetest/main.go (hunk 3 above)
Comment (wking) [reply cmd/runtimetest/main.go]: > Should it necessarily be wrt to a mount in the host namespace?

This is another case where we can work around `runtimetest` limitations by limiting the configs we create in `validation/` to things `runtimetest` can check for ;).

> But for the purpose of this PR, your change is fine.

Yeah, I'm fine making not growing this PR to fix everything ;).  And if folks want me to spin off anything I *am* doing (like padjusting the readonly tests][1]), let me know and I'll try to spin those out into their own PRs.

[1]: https://github.com/opencontainers/runtime-tools/pull/308#discussion_r179808463

File: cmd/runtimetest/main.go (hunk 4 above)
Comment (wking) [reply cmd/runtimetest/main.go]: > `specerror.DevicesAvailable` -> `condition`

Thanks.  I've fixed all of those with 06482da → 031e4c8.

File: cmd/runtimetest/main.go (hunk 5 above)
Comment (wking) [reply cmd/runtimetest/main.go]: > wouldn't it be better to test
>
>    if device.Path == "/dev/console" {

I'm fine with that.  My current check ensures that we *do* perform the rest of these checks if the config JSON explicitly included `/dev/console` in `linux.devices`.  But we don't do that in `validation/`, and I don't expect us to start, so it probably doesn't matter either way.

File: cmd/runtimetest/main.go (hunk 2 above)
Comment (alban) [reply cmd/runtimetest/main.go]: Can it use [WriteFile](https://golang.org/pkg/io/ioutil/#WriteFile)?

File: cmd/runtimetest/main.go (hunk 7 above)
Comment (wking) [reply cmd/runtimetest/main.go]: > Why ignoring the `os.IsNotExist` error?

Because if the path is not there at all, it's certainly not readable ;).  But while that makes some sense for `maskedPaths`, runtimes should probably not go that far with `readonlyPaths`.  I've dropped it from the `readonlyPaths` checker in 031e4c8 → 5e395e8.

File: cmd/runtimetest/main.go (hunk 8 above)
Comment (wking) [reply cmd/runtimetest/main.go]: > why the `if`? The condition is already in the `c.harness.Ok()` below.

A sloppy copy/paste from the non-fatal readable test a few lines up.  Fixed by dropping this `if` in 5e395e8 → 1e67d33.

File: cmd/runtimetest/main.go (hunk 2 above)
Comment (wking) [reply cmd/runtimetest/main.go]: > Can it use WriteFile?

Good idea.  I've updated to use `WriteFile` for the whole PR with 7ecd679 → 92de75c.

**Comment by wking** [comment]
I've fixed a few typos that had snuck in with my recent rerolls, and [Travis is happy again][1].  Also, in case folks are interested, here's an example of how node-tap renders the `runtimetest` output if I launch it with a random `config.json` from my host as an unprivileged user (without bothering to actually use a runtime or launch a container ;):

```console
$ tap ./runtimetest
./runtimetest ..................................... 199/275
  not ok has expected working directory
    actual: /…/src/github.com/opencontainers/runtime-tools
    expected: /
  
  not ok mounts[1] (/tmp) found
  not ok mounts[2] (/dev) found in order
    config:
      destination: /dev
      type: devtmpfs
      source: devtmpfs
    earlier:
      config:
        destination: /tmp
        type: tmpfs
        source: none
      indexConfig: 0
      indexSystem: 18
    indexConfig: 2
    indexSystem: 0
    level: MUST
    reference: 'https://github.com/opencontainers/runtime-spec/blob/v1.1.0/config.md#mounts'
  
  not ok has expected user ID
    actual: 1000
    expected: 1
  
  not ok has expected group ID
    actual: 1000
    expected: 1
  
  not ok unexpected bounding capability CAP_CHOWN not set
  not ok unexpected bounding capability CAP_DAC_OVERRIDE not set
  not ok unexpected bounding capability CAP_DAC_READ_SEARCH not set
  not ok unexpected bounding capability CAP_FOWNER not set
  not ok unexpected bounding capability CAP_FSETID not set
  not ok unexpected bounding capability CAP_KILL not set
  not ok unexpected bounding capability CAP_SETGID not set
  not ok unexpected bounding capability CAP_SETUID not set
  not ok unexpected bounding capability CAP_SETPCAP not set
  not ok unexpected bounding capability CAP_LINUX_IMMUTABLE not set
  not ok unexpected bounding capability CAP_NET_BIND_SERVICE not set
  not ok unexpected bounding capability CAP_NET_BROADCAST not set
  not ok unexpected bounding capability CAP_NET_ADMIN not set
  not ok unexpected bounding capability CAP_NET_RAW not set
  not ok unexpected bounding capability CAP_IPC_LOCK not set
  not ok unexpected bounding capability CAP_IPC_OWNER not set
  not ok unexpected bounding capability CAP_SYS_MODULE not set
  not ok unexpected bounding capability CAP_SYS_RAWIO not set
  not ok unexpected bounding capability CAP_SYS_CHROOT not set
  not ok unexpected bounding capability CAP_SYS_PTRACE not set
  not ok unexpected bounding capability CAP_SYS_PACCT not set
  not ok unexpected bounding capability CAP_SYS_ADMIN not set
  not ok unexpected bounding capability CAP_SYS_BOOT not set
  not ok unexpected bounding capability CAP_SYS_NICE not set
  not ok unexpected bounding capability CAP_SYS_RESOURCE not set
  not ok unexpected bounding capability CAP_SYS_TIME not set
  not ok unexpected bounding capability CAP_SYS_TTY_CONFIG not set
  not ok unexpected bounding capability CAP_MKNOD not set
  not ok unexpected bounding capability CAP_LEASE not set
  not ok unexpected bounding capability CAP_AUDIT_WRITE not set
  not ok unexpected bounding capability CAP_AUDIT_CONTROL not set
  not ok unexpected bounding capability CAP_SETFCAP not set
  not ok unexpected bounding capability CAP_MAC_OVERRIDE not set
  not ok unexpected bounding capability CAP_MAC_ADMIN not set
  not ok unexpected bounding capability CAP_SYSLOG not set
  not ok unexpected bounding capability CAP_WAKE_ALARM not set
  not ok unexpected bounding capability CAP_BLOCK_SUSPEND not set
  not ok unexpected bounding capability CAP_AUDIT_READ not set
  not ok has expected process argument 0
    actual: ./runtimetest
    expected: sh
    index: 0

  Skipped: 32
     root.readonly is false but the root filesystem is still not writable
     hostname not set
     /dev/null (default device) has unconfigured permissions
     /dev/null (default device) has an unconfigured user ID
     /dev/null (default device) has an unconfigured group ID
     /dev/zero (default device) has unconfigured permissions
     /dev/zero (default device) has an unconfigured user ID
     /dev/zero (default device) has an unconfigured group ID
     /dev/full (default device) has unconfigured permissions
     /dev/full (default device) has an unconfigured user ID
     /dev/full (default device) has an unconfigured group ID
     /dev/random (default device) has unconfigured permissions
     /dev/random (default device) has an unconfigured user ID
     /dev/random (default device) has an unconfigured group ID
     /dev/urandom (default device) has unconfigured permissions
     /dev/urandom (default device) has an unconfigured user ID
     /dev/urandom (default device) has an unconfigured group ID
     /dev/tty (default device) has unconfigured permissions
     /dev/tty (default device) has an unconfigured user ID
     /dev/tty (default device) has an unconfigured group ID
     /dev/ptmx (default device) has unconfigured permissions
     /dev/ptmx (default device) has an unconfigured user ID
     /dev/ptmx (default device) has an unconfigured group ID
     linux.devices is not set
     linux.maskedPaths not set
     process.oomScoreAdj not set
     linux.seccomp not set
     linux.readonlyPaths not set
     linux.rootfsPropagation not set
     linux.sysctl not set
     linux.uidMappings not set
     linux.gidMappings not set

total ............................................. 199/275
  

  199 passing (102.442ms)
  32 pending
  44 failing

```

[1]: https://travis-ci.org/opencontainers/runtime-tools/builds/363247555?utm_source=github_status&utm_medium=notification

**Comment by q384566678** [comment]
LGTM
